### PR TITLE
Improve impossible type checker for void-returning functions

### PR DESCRIPTION
**Pull request by rvanvelzen**
Fix phpstan/phpstan#8169

### CODE REVIEW

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Super-interesting! I'd probably put this behind a bleedingEdge toggle.

The `@phpstan-assert` functionality is enabled for everyone because it can be considered just a smarter type inference, but this bit feels a little bit extra and can result in some extra errors, so bleedingEdge would feel safer for me.

File: conf/config.neon [hunk 1]
```diff
@@ -51,6 +51,7 @@ parameters:
 		curlSetOptTypes: false
 		listType: false
 		missingMagicSerializationRule: false
+		nullContextForVoidReturningFunctions: true
```
Comment (ondrejmirtes) [inline conf/config.neon]: false needs to be here

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by herndlm** [comment]
wow you were fast 🎉
but is this safe? e.g. can it be assumed that `@phpstan-assert string $var` internally basically does `is_string()`? I'm just afraid that it could cause false-positives when people use such an assert for something like `assertEmail()` basically.

**Comment by ondrejmirtes** [comment]
@herndlm Yes, Richard expressed this concern earlier but I said to try it. That's why it's in bleedingEdge now only.

**Comment by ondrejmirtes** [comment]
@rvanvelzen I think this one is ready to merge. I really appreciate all the work you do, but I feel like you're overusing the Draft feature a bit :) I respect it and I wouldn't merge a PR in draft state, but sometimes I'm eager to merge something because to my eyes it seems totally fine, so don't hesitate to mark things as ready sooner. We can always fix this in the future.

Another problem is that GitHub doesn't notify me when something's marked as ready so I added a workflow that adds a bot comment to such PRs: https://github.com/phpstan/phpstan-src/commit/e3b97716b77a7b04319c19cda258b7f4f30b7304

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**